src/configured_subscribed_notifications.c

```diff
@@ -0,0 +1,520 @@
+    } else {
+        csn_data->stream_subtree_filter = NULL;
+    }
+    csn_data->stream_xpath_filter = stream_xpath_filter ? strdup(stream_xpath_filter) : NULL;
+    csn_data->stream = strdup(stream);
+
+    if (local_address)
+        csn_data->cb_arg.recv_info.local_address = strdup(local_address);
+
+    if (interface)
+        csn_data->cb_arg.recv_info.interface = strdup(interface);
+
+    csn_data->replay_start_time = start;
+    if ((stream_filter_name && !csn_data->stream_filter_name) || (stream_subtree_filter && !csn_data->stream_subtree_filter) ||
+            (stream_xpath_filter && !csn_data->stream_xpath_filter) || !csn_data->stream) {
+        rc = SR_ERR_NO_MEMORY;
+        goto cleanup;
+    }
+
+    /* no API to get these data in the callback, so they are accessible directly this way, the lock must always be held */
+    csn_data->cb_arg.csn_data = csn_data;
+    csn_data->cb_arg.nc_sub_id = sub->nc_sub_id;
+
+    if (sub->stop_time.tv_sec) {
+        /* create stop timer */
+        rc = sub_ntf_create_timer(csn_stop_timer_cb,
+                                  &csn_data->cb_arg, 1,
+                                  &csn_data->stop_timer);
+        if (rc != SR_ERR_OK) {
+            goto cleanup;
+        }
+    }
+
+    /* subscribe to sysrepo notifications, cb_arg is managed (freed) by the callback */
+    rc = csn_sr_subscribe(stream, xp, start.tv_sec ? &start : NULL,
+            sub->stop_time.tv_sec ? &sub->stop_time : NULL, &csn_data->cb_arg, ev_sess, &sub->sub_ids, &sub_id_count);
+    ATOMIC_STORE_RELAXED(sub->sub_id_count, sub_id_count);
+    if (rc != SR_ERR_OK) {
+        goto cleanup;
+    }
+
+    csn_data->state = CONFIGURED_SUBSCRIPTION_STATE_VALID;
+
+cleanup:
+    free(xp);
+    if (rc) {
+        csn_data_destroy(csn_data);
+    }
+    return rc;
+}
+
+int
+csn_config_sub_start_async(sr_session_ctx_t *UNUSED(ev_sess), struct np2srv_sub_ntf *UNUSED(sub))
+{
+    /* resume subscription thread */
+    sr_subscription_thread_resume(np2srv.sr_cfg_notif_sub);
+
+    return 0;
+}
+
+void
+csn_terminate_async(struct np2srv_sub_ntf *sub)
+{
+    struct csn_data_t *csn_data = sub->data;
+    struct itimerspec tspec = {0};
+
+    if (csn_data->stop_timer) {
+        const struct ly_ctx *ly_ctx;
+        struct lyd_node *ly_ntf;
+        char buf[26];
+
+        timer_settime(csn_data->stop_timer, TIMER_ABSTIME, &tspec, NULL);
+
+        ly_ctx = sr_acquire_context(np2srv.sr_conn);
+        sr_release_context(np2srv.sr_conn);
+
+        sprintf(buf, "%" PRIu32, sub->nc_sub_id);
+        lyd_new_path(NULL, ly_ctx, "/ietf-subscribed-notifications:subscription-completed/id",
+                     buf, 0, &ly_ntf);
+        csn_send_notif(&csn_data->cb_arg.recv_info, sub->nc_sub_id,
+                       np_gettimespec(1), &ly_ntf, 1);
+    }
+}
+
+void
+csn_data_destroy(void *data)
+{
+    struct csn_data_t *csn_data = data;
+    uint32_t i;
+
+    if (csn_data) {
```

Comment:
Minor improvement, I like to write similar functions as `if (!csn_data) { return; }` and then continue with the code without additional indentation.